TIỀN XỬ LÝ DỮ LIỆU

IMPORT THƯ VIỆN

In [8]:
import pandas as pd
import unicodedata
from underthesea import text_normalize, word_tokenize
import regex as re
import emot

ĐỌC DỮ LIỆU

In [9]:
df = pd.read_csv('data/review.csv')
df

id               title  \
0      19906519     Cực kì hài lòng   
1      19968834            Hài lòng   
2      19943636  Rất không hài lòng   
3      19919040     Cực kì hài lòng   
4      19963835            Hài lòng   
...         ...                 ...   
26227  19939226     Cực kì hài lòng   
26228  19931668     Cực kì hài lòng   
26229   9198319     Cực kì hài lòng   
26230  19943620     Cực kì hài lòng   
26231  19900164     Cực kì hài lòng   

                                                 content  thank_count  \
0      Sau khi phản ánh việc tác giả đặt tên sách trù...            8   
1      Tạm thời thấy ổn.Đọc xong sẽ  review sau.hj.Yh...            0   
2      quyển sách với tựa đề làm hiểu lầm cái khách h...            0   
3                     Giao hàng nhanh, đóng gói cẩn thận            0   
4      Tạm được, phù hợp những bạn chưa đọc nhiều sác...            0   
...                                                  ...          ...   
26227                                                NaN            0   
26228                                                NaN            0   
26229  Chất lượng sách tốt, giấy tốt, Tiki giao hàng ...            0   
26230                                                NaN            0   
26231                                                NaN            0   

       customer_id  rating  product_id  
0         13450478       5   275025208  
1         19821878       4   275025208  
2         14317039       1   275025208  
3         13713637       5   275025208  
4          7159313       4   275025208  
...            ...     ...         ...  
26227     13696065       5    73316970  
26228     29245947       5    73316970  
26229      7562223       5    73316970  
26230     12028298       5   274909433  
26231       135809       5   274909433  

[26232 rows x 7 columns]

XÓA DỮ LIỆU BỊ RỖNG

Tính toán data rỗng theo rating

In [10]:
def tim_data_rong(df):
    df_isnull = df[df['content'].isnull()]
    df_count = df_isnull['rating'].value_counts()
    return df_count

In [11]:
tim_data_rong(df)

rating
5    8203
4     579
3       2
1       1
Name: count, dtype: int64

Nhận xét: Phần lớn các comment chỉ đánh giá bằng rating và không nói gì thêm nằm ở mức số 5, tức là đại đa số khách hàng hài lòng và không nói gì thêm, ta có thể bỏ quan NaN values để dữ liệu cần bằng hơn.

In [12]:
def xoa_data_rong(df):
    df.dropna(subset=['content'], inplace=True)
    df.reset_index(drop=True, inplace=True)
    return df

In [13]:
xoa_data_rong(df)

id               title  \
0      19906519     Cực kì hài lòng   
1      19968834            Hài lòng   
2      19943636  Rất không hài lòng   
3      19919040     Cực kì hài lòng   
4      19963835            Hài lòng   
...         ...                 ...   
17442  17345116     Cực kì hài lòng   
17443   8770862     Cực kì hài lòng   
17444  17586361     Cực kì hài lòng   
17445  14238907     Cực kì hài lòng   
17446   9198319     Cực kì hài lòng   

                                                 content  thank_count  \
0      Sau khi phản ánh việc tác giả đặt tên sách trù...            8   
1      Tạm thời thấy ổn.Đọc xong sẽ  review sau.hj.Yh...            0   
2      quyển sách với tựa đề làm hiểu lầm cái khách h...            0   
3                     Giao hàng nhanh, đóng gói cẩn thận            0   
4      Tạm được, phù hợp những bạn chưa đọc nhiều sác...            0   
...                                                  ...          ...   
17442  Giấy vàng không bị lóa mắt\r\nMực in đều\r\nĐó...            0   
17443  SÁCH HAY NÊN ĐỌC\nCuốn sách làm thay đổi tư du...            1   
17444  Giao hàng nhanh,shipper thân thiện và nhiệt tì...            0   
17445  Nghe về tác giả nhiều rồi nên hôm nay mua mấy ...            0   
17446  Chất lượng sách tốt, giấy tốt, Tiki giao hàng ...            0   

       customer_id  rating  product_id  
0         13450478       5   275025208  
1         19821878       4   275025208  
2         14317039       1   275025208  
3         13713637       5   275025208  
4          7159313       4   275025208  
...            ...     ...         ...  
17442     16638755       5    73316970  
17443      5330292       5    73316970  
17444     12472234       5    73316970  
17445      1110677       5    73316970  
17446      7562223       5    73316970  

[17447 rows x 7 columns]

XÓA DỮ LIỆU TRÙNG LẶP

In [14]:
def xoa_dup(df):
    df.drop_duplicates(subset=['content'],inplace=True)
    df.reset_index(drop=True, inplace=True)
    return df

In [15]:
xoa_dup(df)

id               title  \
0      19906519     Cực kì hài lòng   
1      19968834            Hài lòng   
2      19943636  Rất không hài lòng   
3      19919040     Cực kì hài lòng   
4      19963835            Hài lòng   
...         ...                 ...   
16079  17345116     Cực kì hài lòng   
16080   8770862     Cực kì hài lòng   
16081  17586361     Cực kì hài lòng   
16082  14238907     Cực kì hài lòng   
16083   9198319     Cực kì hài lòng   

                                                 content  thank_count  \
0      Sau khi phản ánh việc tác giả đặt tên sách trù...            8   
1      Tạm thời thấy ổn.Đọc xong sẽ  review sau.hj.Yh...            0   
2      quyển sách với tựa đề làm hiểu lầm cái khách h...            0   
3                     Giao hàng nhanh, đóng gói cẩn thận            0   
4      Tạm được, phù hợp những bạn chưa đọc nhiều sác...            0   
...                                                  ...          ...   
16079  Giấy vàng không bị lóa mắt\r\nMực in đều\r\nĐó...            0   
16080  SÁCH HAY NÊN ĐỌC\nCuốn sách làm thay đổi tư du...            1   
16081  Giao hàng nhanh,shipper thân thiện và nhiệt tì...            0   
16082  Nghe về tác giả nhiều rồi nên hôm nay mua mấy ...            0   
16083  Chất lượng sách tốt, giấy tốt, Tiki giao hàng ...            0   

       customer_id  rating  product_id  
0         13450478       5   275025208  
1         19821878       4   275025208  
2         14317039       1   275025208  
3         13713637       5   275025208  
4          7159313       4   275025208  
...            ...     ...         ...  
16079     16638755       5    73316970  
16080      5330292       5    73316970  
16081     12472234       5    73316970  
16082      1110677       5    73316970  
16083      7562223       5    73316970  

[16084 rows x 7 columns]

Dữ liệu sau khi bị xóa trùng lặp và xóa rỗng thì còn 16084 dòng

XÓA EMOJI VÀ EMOTICON

In [16]:
#Khởi tạo đối tượng emot
emot_obj = emot.core.emot()

In [17]:
def xoa_emoji_va_emoticon(text):
    emojis = emot_obj.emoji(text)
    emoticons = emot_obj.emoticons(text)
    for i in emojis['value'] + emoticons['value']:
        for punc in ['(', ')', '.', '+', '*']:
            i = i.replace(punc, '\\' + punc)
        pattern = r'\s' + i + r'\s'
        text = re.sub(pattern, ' ', text)
    return text

CHUYỂN TEENCODE THÀNH TỪ ĐÚNG

Đọc file teencode

In [18]:
with open('files/teencode.txt', 'r', encoding='utf8') as file:
    teencode_list = file.read().split('\n')
    teencode_dict = {}
    for line in teencode_list:
        key, value = line.split('\t')
        teencode_dict[key] = str(value)
file.close()

Chuyển teencode thành từ đúng

In [19]:
def chuyen_teencode(text):
    text = ' '.join(teencode_dict[word] if word in teencode_dict else word for word in text.split())
    return text

XÓA CÁC TỪ KHÔNG ĐÚNG TRONG TIẾNG VIỆT

Đọc file có sẵn chứa các từ sai

In [20]:
with open('files/wrong-word.txt', 'r', encoding='utf8') as file:
    wrongword_list = file.read().split('\n')
file.close()

Xóa từ bị sai

In [21]:
def xoa_wrongword(text):
    text = ' '.join('' if word in wrongword_list else word for word in text.split())
    return text

CHUẨN HÓA DẤU CÂU VÀ LOẠI BỎ CÁC THÀNH PHẦN KHÔNG CẦN THIẾT

In [22]:
def xoa_dau_cau_va_so(text):
    pattern = r'(?i)\b[a-záàảãạăắằẳẵặâấầẩẫậéèẻẽẹêếềểễệóòỏõọôốồổỗộơớờởỡợíìỉĩịúùủũụưứừửữựýỳỷỹỵđ]+\b'
    text = ' '.join(re.findall(pattern, text))
    text = text.lower()
    text = re.sub(r'\s+', ' ', text).strip()  
    return text

CHUẨN HÓA UNICODE TIẾNG VIỆT

Có 2 loaị Unicode được sử dụng phổ biến là Unicode tổ hợp và Unicode dựng sẵn. Trong bài này, tôi chọn chuẩn hóa về Unicode dựng sẵn thì loại này được sử dụng phổ biến hơn.

In [23]:
def loaddicchar():
    uniChars = "àáảãạâầấẩẫậăằắẳẵặèéẻẽẹêềếểễệđìíỉĩịòóỏõọôồốổỗộơờớởỡợùúủũụưừứửữựỳýỷỹỵÀÁẢÃẠÂẦẤẨẪẬĂẰẮẲẴẶÈÉẺẼẸÊỀẾỂỄỆĐÌÍỈĨỊÒÓỎÕỌÔỒỐỔỖỘƠỜỚỞỠỢÙÚỦŨỤƯỪỨỬỮỰỲÝỶỸỴÂĂĐÔƠƯ"
    unsignChars = "aaaaaaaaaaaaaaaaaeeeeeeeeeeediiiiiooooooooooooooooouuuuuuuuuuuyyyyyAAAAAAAAAAAAAAAAAEEEEEEEEEEEDIIIOOOOOOOOOOOOOOOOOOOUUUUUUUUUUUYYYYYAADOOU"
    dic = {}
    char1252 = 'à|á|ả|ã|ạ|ầ|ấ|ẩ|ẫ|ậ|ằ|ắ|ẳ|ẵ|ặ|è|é|ẻ|ẽ|ẹ|ề|ế|ể|ễ|ệ|ì|í|ỉ|ĩ|ị|ò|ó|ỏ|õ|ọ|ồ|ố|ổ|ỗ|ộ|ờ|ớ|ở|ỡ|ợ|ù|ú|ủ|ũ|ụ|ừ|ứ|ử|ữ|ự|ỳ|ý|ỷ|ỹ|ỵ|À|Á|Ả|Ã|Ạ|Ầ|Ấ|Ẩ|Ẫ|Ậ|Ằ|Ắ|Ẳ|Ẵ|Ặ|È|É|Ẻ|Ẽ|Ẹ|Ề|Ế|Ể|Ễ|Ệ|Ì|Í|Ỉ|Ĩ|Ị|Ò|Ó|Ỏ|Õ|Ọ|Ồ|Ố|Ổ|Ỗ|Ộ|Ờ|Ớ|Ở|Ỡ|Ợ|Ù|Ú|Ủ|Ũ|Ụ|Ừ|Ứ|Ử|Ữ|Ự|Ỳ|Ý|Ỷ|Ỹ|Ỵ'.split('|')
    charutf8 = "à|á|ả|ã|ạ|ầ|ấ|ẩ|ẫ|ậ|ằ|ắ|ẳ|ẵ|ặ|è|é|ẻ|ẽ|ẹ|ề|ế|ể|ễ|ệ|ì|í|ỉ|ĩ|ị|ò|ó|ỏ|õ|ọ|ồ|ố|ổ|ỗ|ộ|ờ|ớ|ở|ỡ|ợ|ù|ú|ủ|ũ|ụ|ừ|ứ|ử|ữ|ự|ỳ|ý|ỷ|ỹ|ỵ|À|Á|Ả|Ã|Ạ|Ầ|Ấ|Ẩ|Ẫ|Ậ|Ằ|Ắ|Ẳ|Ẵ|Ặ|È|É|Ẻ|Ẽ|Ẹ|Ề|Ế|Ể|Ễ|Ệ|Ì|Í|Ỉ|Ĩ|Ị|Ò|Ó|Ỏ|Õ|Ọ|Ồ|Ố|Ổ|Ỗ|Ộ|Ờ|Ớ|Ở|Ỡ|Ợ|Ù|Ú|Ủ|Ũ|Ụ|Ừ|Ứ|Ử|Ữ|Ự|Ỳ|Ý|Ỷ|Ỹ|Ỵ".split('|')
    for i in range(len(char1252)):
        dic[char1252[i]] = charutf8[i]
    return dic
         
def chuyen_unicode(text):
    dicchar = loaddicchar()
    return re.sub(
        r'à|á|ả|ã|ạ|ầ|ấ|ẩ|ẫ|ậ|ằ|ắ|ẳ|ẵ|ặ|è|é|ẻ|ẽ|ẹ|ề|ế|ể|ễ|ệ|ì|í|ỉ|ĩ|ị|ò|ó|ỏ|õ|ọ|ồ|ố|ổ|ỗ|ộ|ờ|ớ|ở|ỡ|ợ|ù|ú|ủ|ũ|ụ|ừ|ứ|ử|ữ|ự|ỳ|ý|ỷ|ỹ|ỵ|À|Á|Ả|Ã|Ạ|Ầ|Ấ|Ẩ|Ẫ|Ậ|Ằ|Ắ|Ẳ|Ẵ|Ặ|È|É|Ẻ|Ẽ|Ẹ|Ề|Ế|Ể|Ễ|Ệ|Ì|Í|Ỉ|Ĩ|Ị|Ò|Ó|Ỏ|Õ|Ọ|Ồ|Ố|Ổ|Ỗ|Ộ|Ờ|Ớ|Ở|Ỡ|Ợ|Ù|Ú|Ủ|Ũ|Ụ|Ừ|Ứ|Ử|Ữ|Ự|Ỳ|Ý|Ỷ|Ỹ|Ỵ',
        lambda x: dicchar[x.group()], text)

CHUẨN HÓA DẤU CÂU TRONG TIẾNG VIỆT

Có 2 kiểu gõ dấu ở Tiếng Việt, ví dụ như là : òa hoặc oà (ta gọi lần lượt là chuẩn 1 và 2). Mặc dù kiểu gõ chữ sau ít phổ biến hơn tuy nhiên vẫn cần phải chuẩn hóa tránh việc một số văn bản vẫn sử dụng kiểu gõ dấu thứ 2.

In [24]:
def chuan_hoa_dau_tieng_viet(text):
    text = text_normalize(text)
    return text

TÁCH TỪ

In [25]:
def tach_tu_tieng_viet(text):
    text = word_tokenize(text, format="text")
    return text

XÓA STOPWORDS

Đọc file chứa các từ stopword trong Tiếng Việt

In [26]:
with open('files/vietnamese-stopwords.txt', 'r', encoding='utf8') as file:
    stopword_list = file.read().split('\n')
file.close()

Xóa stopword

In [27]:
def xoa_stopword(text):
    text = ' '.join('' if word in stopword_list else word for word in text.split())
    text = re.sub(r'\s+', ' ', text).strip()
    return text

TIỀN XỬ LÝ DỮ LIỆU

In [28]:
def tien_xu_li(text):
    text = xoa_emoji_va_emoticon(text)
    text = chuyen_teencode(text)
    text = xoa_wrongword(text)
    text = xoa_dau_cau_va_so(text)
    text = chuyen_unicode(text)
    text = chuan_hoa_dau_tieng_viet(text)
    text = tach_tu_tieng_viet(text)
    text = xoa_stopword(text)
    return text

In [29]:
text_list = df['content'].values.tolist()
for text in text_list:
    processed_text = tien_xu_li(text)

In [30]:
df['clean_content'] = df['content'].apply(tien_xu_li)
df.shape

(16084, 8)

XÓA DATA RỖNG VÀ LƯU DỮ LIỆU ĐÃ XỬ LÝ VÀO FILE CSV

Có những đánh giá có nội dung trùng lặp nhau do khách hàng mua sách theo combo, đánh giá theo đơn hàng nên sẽ có tình trạng cùng 1 content nhưng lại có 2 review id khác nhau.

In [31]:
def luu_data(df):   
    df.dropna(subset= ['clean_content'], inplace=True)
    df.drop_duplicates(subset= ['clean_content'], inplace=True)
    df.reset_index(drop=True,  inplace=True)
    df.to_csv('data/clean_review.csv',index=False, encoding='utf8')
    return df

In [32]:
luu_data(df)

id               title  \
0      19906519     Cực kì hài lòng   
1      19968834            Hài lòng   
2      19943636  Rất không hài lòng   
3      19919040     Cực kì hài lòng   
4      19963835            Hài lòng   
...         ...                 ...   
15398  17345116     Cực kì hài lòng   
15399   8770862     Cực kì hài lòng   
15400  17586361     Cực kì hài lòng   
15401  14238907     Cực kì hài lòng   
15402   9198319     Cực kì hài lòng   

                                                 content  thank_count  \
0      Sau khi phản ánh việc tác giả đặt tên sách trù...            8   
1      Tạm thời thấy ổn.Đọc xong sẽ  review sau.hj.Yh...            0   
2      quyển sách với tựa đề làm hiểu lầm cái khách h...            0   
3                     Giao hàng nhanh, đóng gói cẩn thận            0   
4      Tạm được, phù hợp những bạn chưa đọc nhiều sác...            0   
...                                                  ...          ...   
15398  Giấy vàng không bị lóa mắt\r\nMực in đều\r\nĐó...            0   
15399  SÁCH HAY NÊN ĐỌC\nCuốn sách làm thay đổi tư du...            1   
15400  Giao hàng nhanh,shipper thân thiện và nhiệt tì...            0   
15401  Nghe về tác giả nhiều rồi nên hôm nay mua mấy ...            0   
15402  Chất lượng sách tốt, giấy tốt, Tiki giao hàng ...            0   

       customer_id  rating  product_id  \
0         13450478       5   275025208   
1         19821878       4   275025208   
2         14317039       1   275025208   
3         13713637       5   275025208   
4          7159313       4   275025208   
...            ...     ...         ...   
15398     16638755       5    73316970   
15399      5330292       5    73316970   
15400     12472234       5    73316970   
15401      1110677       5    73316970   
15402      7562223       5    73316970   

                                           clean_content  
0      phản_ánh tác_giả sách_trùng sách nước_ngoài nh...  
1                       tạm_thời ổn đọc review hj yhanks  
2      quyển sách tựa_đề lầm khách_hàng viết không đủ...  
3                            giao hàng đóng_gói cẩn_thận  
4             tạm đọc sách tác_giả nội_sưu_tầm trùng_lặp  
...                                                  ...  
15398  giấy vàng không lóa mắt mực đóng_gói giao hàng...  
15399  sách đọc sách đổi_tư phương_pháp giáo_dục trẻ ...  
15400  giao hàng shipper thân_thiện nhiệt_tình sách c...  
15401       tác_giả hôm mấy đọc thử tiki hàng chất_lượng  
15402  chất_lượng sách tốt giấy tốt tiki hàng hôm hôm...  

[15403 rows x 8 columns]